File: #36_Valid_Sudoku/Solution.java
import java.util.*;

public class Solution {
    public boolean isValidSudoku(char[][] board) {
        if (board == null || board.length == 0) return false;
        BitSet[] rows = new BitSet[9];
        BitSet[] cols = new BitSet[9];
        BitSet[] subs = new BitSet[9];
        for (int i = 0; i < 9; i++) {
        	rows[i] = new BitSet(10);
        }
        for (int i = 0; i < 9; i++) {
        	cols[i] = new BitSet(10);
        }
        for (int i = 0; i < 9; i++) {
        	subs[i] = new BitSet(10);
        }
//      for (BitSet bs : rows) bs = new BitSet(10);
//   	for (BitSet bs : cols) bs = new BitSet(10);
//      for (BitSet bs : subs) bs = new BitSet(10);
		for (int i = 0; i < board.length; i++) {
			for (int j = 0; j < board[0].length; j++) {
				if (board[i][j] == '.') continue;
				int num = board[i][j] - '0';
				if (rows[i].get(num)) return false;
				rows[i].set(num);
				if (cols[j].get(num)) return false;
				cols[j].set(num);
				int k = 3 * (i / 3) + j / 3;
				if (subs[k].get(num)) return false;
				subs[k].set(num);
			}
		}
		return true;
    }

    public static void main(String[] args) {
    	char[][] board = {
    		{'.', '8', '7', '6', '5', '4', '3', '2', '1'},
    		{'2', '.', '.', '.', '.', '.', '.', '.', '.'},
    		{'3', '.', '.', '.', '.', '.', '.', '.', '.'},
    		{'4', '.', '.', '.', '.', '.', '.', '.', '.'},
    		{'5', '.', '.', '.', '.', '.', '.', '.', '.'},
    		{'6', '.', '.', '.', '.', '.', '.', '.', '.'},
    		{'7', '.', '.', '.', '.', '.', '.', '.', '.'},
    		{'8', '.', '.', '.', '.', '.', '.', '.', '.'},
    		{'9', '.', '.', '.', '.', '.', '.', '.', '.'}
    	};
    	Solution sol = new Solution();
    	System.out.println(sol.isValidSudoku(board));
    }
}

File: #36_Valid_Sudoku/Solution.py
class Solution:
	# @param {character[][]} board
	# @return {boolean}
	def isValidSudoku(self, board):
		if not board or not board[0]:
			return False
		row = [[False for j in range(9)] for i in range(9)]
		col = [[False for j in range(9)] for i in range(9)]
		box = [[False for j in range(9)] for i in range(9)]
		for i in range(9):
			for j in range(9):
				if board[i][j] == '.':
					continue
				num = ord(board[i][j]) - '1'
				indexBox = self.findBox(i, j)
				if row[i][num] or col[j][num] or box[indexBox][num]:
					return False
				row[i][num], col[j][num], box[self.findBox(i, j)][num] = True, True, True
		return True

	def findBox(self, i, j):
		return 3 * (i / 3) + j / 3